File: adventofcode/2019/d1/d1.go
package main

import (
	"io/ioutil"
	"strconv"
	"strings"
)

func calcTotalFuel(mass int) int {
	fuel := calcFuel(mass)
	if fuel <= 0 {
		return 0
	}
	return fuel + calcTotalFuel(fuel)
}

func calcFuel(mass int) int {
	return (mass / 3) - 2
}

func castToInts(vs []string) []int {
	arr := make([]int, len(vs))
	for i, v := range vs {
		casted, _ := strconv.Atoi(v)
		arr[i] = casted
	}
	return arr
}

func main() {
	content, _ := ioutil.ReadFile("./input.txt")
	masses := castToInts(strings.Split(string(content), "\n"))

	// p1
	p1sum := 0
	for _, v := range masses {
		p1sum += calcFuel(v)
	}
	println(p1sum)

	// p2
	p2sum := 0
	for _, v := range masses {
		p2sum += calcTotalFuel(v)
	}
	println(p2sum)

}

File: adventofcode/2019/d1/d1.js
const fs = require('fs')

const masses = fs.readFileSync('./input.txt').toString().split("\n").map(Number)

const calcFuel = n => Math.floor(n / 3) - 2;

const calcTotalFuel = n => {
  const fuel = calcFuel(n);
  return fuel <= 0 ? 0 : fuel + calcTotalFuel(fuel);
};

const sum = (a, b) => a + b;

// p1
console.log(masses.map(calcFuel).reduce(sum))

// p2
console.log(masses.map(calcTotalFuel).reduce(sum))
